### PR TITLE
[LTS_07_2022] Refresh for OpenSSL 3 and Ubuntu 22.04 Compatibility

### PR DESCRIPTION
**Pull request by vaavva**
# Checklist
- [x] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-c/blob/main/.github/CONTRIBUTING.md).
- [x] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.
- I submitted this PR against the correct branch: 
  - [ ] This pull-request is submitted against the `main` branch. 
  - [ ] I have merged the latest `main` branch prior to submission and re-merged as needed after I took any feedback.
  - [ ] I have squashed my changes into one with a clear description of the change.

# Reference/Link to the issue solved with this PR (if any)

# Description of the problem

# Description of the solution

### CODE REVIEW

**Review by danewalton — APPROVED** [review]

File: build/.vsts-ci.yml [hunk 1]
```diff
@@ -447,6 +450,55 @@ jobs:
   - script: sudo rm -rf $(Agent.BuildDirectory)/*
     displayName: 'Cleanup'
     condition: always()
+- job: ubuntu2204build
+  container: linux-c-ubuntu-2204
+  pool:
+    name: 'sdk-c--ubuntu-18'
+  displayName: 'Ubuntu 22.04 Unit Tests (Debug)'
```
Comment (ericwol-msft) [inline build/.vsts-ci.yml:457]:
```suggestion
  displayName: 'Ubuntu 22.04 (Debug)'
```

File: provisioning_client/tests/common_prov_e2e/prov_valgrind_suppression.supp [hunk 2]
```diff
@@ -297,3 +297,78 @@
...
+  fun:OPENSSL_init_ssl
+  obj:/usr/lib/x86_64-linux-gnu/libcurl.so.4.6.0
+  obj:/usr/lib/x86_64-linux-gnu/libcurl.so.4.6.0
+  fun:HTTPAPI_Init
+  fun:HTTPAPIEX_Init
+  fun:platform_init
+  fun:TestSuiteInitialize
+  fun:RunTests
+  fun:main
+}
```
Comment (ericwol-msft) [inline provisioning_client/tests/common_prov_e2e/prov_valgrind_suppression.supp:374]: newline

File: iothub_client/tests/global_valgrind_suppression.supp [hunk 3]
```diff
@@ -395,4 +395,239 @@
...
+  obj:/usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
+  obj:/usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
+  fun:__pthread_once_slow
+  fun:CRYPTO_THREAD_run_once
+  fun:OPENSSL_init_crypto
+  fun:OPENSSL_init_ssl
+  obj:/usr/lib/x86_64-linux-gnu/libcurl.so.4.6.0
+}
+{
+  <insert_a_suppression_name_here>
```
Comment (ericwol-msft) [inline iothub_client/tests/global_valgrind_suppression.supp:422]: add a suppression name?

File: iothub_client/tests/global_valgrind_suppression.supp [hunk 4]
```diff
@@ -395,4 +395,239 @@
    fun:e2e_d2c_with_svc_fault_ctrl_error_message_callback
    fun:RunTests
    fun:(below main)
+}
+{
+  OpenSSL leak
```
Comment (ericwol-msft) [inline iothub_client/tests/global_valgrind_suppression.supp:400]:
```suggestion
  OpenSSL RSA leak
```

File: iothub_client/tests/global_valgrind_suppression.supp (hunk 3 above)
Comment (ericwol-msft) [inline iothub_client/tests/global_valgrind_suppression.supp:422]: These are all dups. You can have one statement for all of them.

> {
  OpenSSL EC KEY leak 
  Memcheck:Leak
  match-leak-kinds: definite
  fun:malloc
  fun:CRYPTO_zalloc
  fun:EC_KEY_METHOD_new
  obj:*
  obj:*
  obj:/usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
  fun:ENGINE_ctrl_cmd_string
  fun:ENGINE_by_id
  obj:/usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
  fun:CONF_modules_load
  fun:CONF_modules_load_file
  obj:/usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
  obj:/usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
  fun:__pthread_once_slow
  fun:CRYPTO_THREAD_run_once
  fun:OPENSSL_init_crypto
  fun:OPENSSL_init_ssl
  obj:/usr/lib/x86_64-linux-gnu/libcurl.so.4.6.0
  obj:/usr/lib/x86_64-linux-gnu/libcurl.so.4.6.0
  fun:HTTPAPI_Init
  fun:HTTPAPIEX_Init
  fun:platform_init
  ...
}

File: provisioning_client/tests/common_prov_e2e/prov_valgrind_suppression.supp [hunk 5]
```diff
@@ -297,3 +297,78 @@
...
+{
+  <insert_a_suppression_name_here>
```
Comment (ericwol-msft) [inline provisioning_client/tests/common_prov_e2e/prov_valgrind_suppression.supp:349]: same here. use **...** for all stallstacks under platform_init

**Review by ericwol-msft — APPROVED** [review]

File: build/.vsts-ci.yml [hunk 6]
```diff
@@ -447,6 +450,55 @@ jobs:
...
+      gcc --version
+      openssl version
+      curl --version
+      sudo -E env PATH="$PATH" ./build_all/linux/build.sh --run-unittests --provisioning --use-hsmsymmkey --use-hsmsas --use-hsmx509 --use-edge-modules --config Debug
+    displayName: 'Build'
+  - script: |
+      export OPENSSL_ia32cap=0x00000000
+      sudo chmod -R 755 .
+      cd cmake && sudo -E ../build_all/linux/run_tests.sh
+    displayName: "Run Tests"
```
Comment (ericwol-msft) [inline build/.vsts-ci.yml:470]:
```suggestion
    displayName: "Run Unit Tests"
```

File: build/.vsts-ci.yml (hunk 1 above)
Comment (ericwol-msft) [inline build/.vsts-ci.yml:457]:
```suggestion
  displayName: 'Ubuntu 22.04 (Debug)'
```

Comment (danewalton) [reply build/.vsts-ci.yml:457]: All of these changes are from the merged changes in main. Do we want to have a difference of files/changes?

Comment (CIPop) [reply build/.vsts-ci.yml:457]: I advise against making any changes here - this should be just a merge.
We can make the suggested changes in `main` then backport to LTS at al later time.

**Review by CIPop — APPROVED** [review]

**Pull request merged**